### ISSUE TITLE
Widget property on "properties" in json schema causes unexpected behavior.

### ISSUE THREAD

**Issue by coreConvention** [ngx-formly/ngx-formly#2927]
I have a process that uses a config based approach to apply widget properties to an otherwise widget-less Json schema object.

The original schema after my modifications looks like this, some of the first level properties were removed to make it shorter...

```
{
    "schema": "http://json-schema.org/draft-06/schema#",
    "type": "object",
    "title": "data",
    "additionalProperties": false,
    "properties": {
        "subject": {
            "type": "object",
            "title": "Subject Details",
            "properties": {
                "txtFirstName": {
                    "title": "First Name",
                    "description": "Patient First Name",
                    "type": "string",
                    "widget": {
                        "formlyConfig": {
                            "templateOptions": {
                                "fxFlex": 50
                            }
                        },
                        "order": 0
                    }
                },
                "txtLastName": {
                    "title": "Last Name",
                    "description": "Patient Last Name",
                    "type": "string",
                    "widget": {
                        "formlyConfig": {
                            "templateOptions": {
                                "fxFlex": 50
                            }
                        },
                        "order": 0
                    }
                },
                "dtBirthDate": {
                    "title": "Date of Birth",
                    "description": "Patient Date of Birth",
                    "type": "string",
                    "format": "date",
                    "widget": {
                        "formlyConfig": {
                            "className": "test",
                            "templateOptions": {
                                "fxFlex": 25
                            }
                        },
                        "order": 0
                    }
                },
                "txtCity": {
                    "title": "City",
                    "description": "Patient City",
                    "type": "string",
                    "widget": {
                        "formlyConfig": {
                            "templateOptions": {
                                "fxFlex": 50
                            }
                        },
                        "order": 0
                    }
                },
                "txtState": {
                    "title": "State",
                    "description": "Patient State",
                    "type": "string",
                    "widget": {
                        "formlyConfig": {
                            "templateOptions": {
                                "fxFlex": 25,
                                "options": [
                                    {
                                        "label": "Alabama",
                                        "value": "AL"
                                    },
                                    {
                                        "label": "Alaska",
                                        "value": "AK"
                                    }
                                ]
                            },
                            "key": "state",
                            "type": "select"
                        },
                        "order": 0
                    }
                },
                "selGender": {
                    "title": "Gender",
                    "description": "Patient Gender",
                    "type": "string",
                    "enum": [
                        "Male",
                        "Female",
                        "Other"
                    ]
                }
            },
            "required": [
                "txtFirstName",
                "txtLastName"
            ],
            "additionalProperties": false,
            "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. <a href=\"#\">Excepteur</a> sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "widget": {
                "formlyConfig": {
                    "type": "flex-layout",
                    "templateOptions": {
                        "fxLayout": "row wrap"
                    }
                }
            }
        }
    },
    "widget": {
        "formlyConfig": {
            "type": "tabs"
        }
    }
    
```
Note the widget properties applied in the first level property "subject"
    
The goal of this was to apply a custom flex wrapper (flex-layout) to each section "subject" in this case via the widget property. However, when this runs through toFieldConfig a couple weird things happen. First in the resultant fields the following line in formly-json-schema.service.ts. Ends up stuffing the entirety of the first field group in the list into the fieldGroup of every top level property (siblings of "subject"). As its walking back up the stack during recursion,  the field property in this case appears to be the entire fieldGroup that was just generated for 2nd tier properties of this first trier property, if that makes sense?

    Line 263 in formly-json-schema.service.ts
    // map in possible formlyConfig options from the widget property
    if (schema['widget'] && schema['widget'].formlyConfig) {
      field = reverseDeepMerge(schema['widget'].formlyConfig, field);
    }
    
 The second weird thing is that the original schema passed in to toFieldConfig is  changed by reference. Which oddly enough ends up stuffing the widget.formlyConfig.fieldGroup into every widget property in every sibling of subject. This ends up essentially overriding the properties from the schema.
 
Are widget properties supported like this? I tried to make a stackblitz to show this behavior, and maybe I am using the wrong template for this job. But I think if you take the above json schema and run it thorough your function you will see what I am talking about. Anyway, this library has been fantastic other than this one weird issue. Nevertheless I appreciate all the hard work you all have done.

**Comment by aitboudad** [comment]
Hi, could you please show me an example of expected vs current behavior of those two issues that you've reported (still not clear yet 😅)

**Comment by coreConvention** [comment]
Yea I would be happy to if I could! hah! I tried to adapt one of the stackblitz projects to this but has having issues reproducing it. I did figure out a work around on my end though. It has to do with the way I was generating the object that I was inserting into the widget on each schema property. It was somehow taking it by reference, and applying it over every top level property (#/properties) on the schema that I wasn't explicitly assigning. I am doing this using a "Json UI Schema" like config, that uses scopes to apply field configs to the schema programmatically before calling your function. I got around this by deep cloning every formlyConfig property that I was assigning with each iteration.

That said, what the issue comes down to is that the toFieldConfig function actually mutates the schema you pass to it as it recurses though, and for whatever reason when you get to the line I posted above, in my case, the field its trying to merge there ends up being the field config that was assigned to the child of the previous inner call.

I have all this stuff rolled into a lib using ngModule that I can probably put up in a github repo for you to look at, not asking you to debug my code. But understanding how its working might help you understand what's going on with the toFieldConfig function. Perhaps  I can do a video capture of it happening from my machine it that would help. Its just super difficult to reproduce this without the entire app being in place holistically, and given this is in one of my clients projects I cant just post the whole thing. I might try to clone this repo and strip everything out to bare minimum and see if I can get that to you. Sorry for the inconvenience.
